File: products/sdlc-cc/platform/observability/logging/go/structured_logger.go
package logging

import (
	"context"
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/sirupsen/logrus"
)

// LogLevel represents the severity of a log entry
type LogLevel string

const (
	LevelDebug LogLevel = "debug"
	LevelInfo  LogLevel = "info"
	LevelWarn  LogLevel = "warn"
	LevelError LogLevel = "error"
	LevelFatal LogLevel = "fatal"
	LevelPanic LogLevel = "panic"
)

// LogEntry represents a structured log entry
type LogEntry struct {
	Timestamp     time.Time              `json:"timestamp"`
	Level         LogLevel               `json:"level"`
	Message       string                 `json:"message"`
	Service       string                 `json:"service"`
	Version       string                 `json:"version,omitempty"`
	TraceID       string                 `json:"trace_id,omitempty"`
	SpanID        string                 `json:"span_id,omitempty"`
	CorrelationID string                 `json:"correlation_id,omitempty"`
	UserID        string                 `json:"user_id,omitempty"`
	TenantID      string                 `json:"tenant_id,omitempty"`
	RequestID     string                 `json:"request_id,omitempty"`
	SessionID     string                 `json:"session_id,omitempty"`
	Source        string                 `json:"source,omitempty"`
	Function      string                 `json:"function,omitempty"`
	File          string                 `json:"file,omitempty"`
	Line          int                    `json:"line,omitempty"`
	Duration      time.Duration          `json:"duration,omitempty"`
	Error         *ErrorInfo             `json:"error,omitempty"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
	Tags          []string               `json:"tags,omitempty"`
}

// ErrorInfo contains detailed error information
type ErrorInfo struct {
	Type       string                 `json:"type"`
	Message    string                 `json:"message"`
	StackTrace string                 `json:"stack_trace,omitempty"`
	Cause      string                 `json:"cause,omitempty"`
	Details    map[string]interface{} `json:"details,omitempty"`
}

// LoggerConfig holds configuration for the structured logger
type LoggerConfig struct {
	Service       string   `json:"service" yaml:"service"`
	Version       string   `json:"version" yaml:"version"`
	Environment   string   `json:"environment" yaml:"environment"`
	Level         string   `json:"level" yaml:"level"`
	Format        string   `json:"format" yaml:"format"` // json, text
	Output        string   `json:"output" yaml:"output"` // stdout, stderr, file
	File          string   `json:"file" yaml:"file"`
	MaxSize       int      `json:"max_size" yaml:"max_size"` // MB
	MaxBackups    int      `json:"max_backups" yaml:"max_backups"`
	MaxAge        int      `json:"max_age" yaml:"max_age"` // days
	Compress      bool     `json:"compress" yaml:"compress"`
	EnableTracing bool     `json:"enable_tracing" yaml:"enable_tracing"`
	RedactFields  []string `json:"redact_fields" yaml:"redact_fields"`
}

// DefaultLoggerConfig returns default logger configuration
func DefaultLoggerConfig() *LoggerConfig {
	return &LoggerConfig{
		Service:       "sdlc-platform",
		Version:       "1.0.0",
		Environment:   "development",
		Level:         "info",
		Format:        "json",
		Output:        "stdout",
		MaxSize:       100,
		MaxBackups:    5,
		MaxAge:        30,
		Compress:      true,
		EnableTracing: true,
		RedactFields:  []string{"password", "token", "secret", "key", "auth", "credential"},
	}
}

// StructuredLogger implements structured logging with correlation support
type StructuredLogger struct {
	config   *LoggerConfig
	logger   *logrus.Logger
	redactor *FieldRedactor
}

// FieldRedactor handles sensitive data redaction
type FieldRedactor struct {
	fields map[string]bool
}

// NewFieldRedactor creates a new field redactor
func NewFieldRedactor(fields []string) *FieldRedactor {
	redactor := &FieldRedactor{
		fields: make(map[string]bool),
	}
	for _, field := range fields {
		redactor.fields[strings.ToLower(field)] = true
	}
	return redactor
}

// Redact recursively redacts sensitive fields from a map
func (r *FieldRedactor) Redact(data interface{}) interface{} {
	switch v := data.(type) {
	case map[string]interface{}:
		redacted := make(map[string]interface{})
		for key, value := range v {
			if r.isSensitive(key) {
				redacted[key] = "***REDACTED***"
			} else {
				redacted[key] = r.Redact(value)
			}
		}
		return redacted
	case []interface{}:
		redacted := make([]interface{}, len(v))
		for i, item := range v {
			redacted[i] = r.Redact(item)
		}
		return redacted
	default:
		return v
	}
}

func (r *FieldRedactor) isSensitive(field string) bool {
	lowerField := strings.ToLower(field)
	for sensitive := range r.fields {
		if strings.Contains(lowerField, sensitive) {
			return true
		}
	}
	return false
}

// NewStructuredLogger creates a new structured logger
func NewStructuredLogger(config *LoggerConfig) (*StructuredLogger, error) {
	if config == nil {
		config = DefaultLoggerConfig()
	}

	logger := logrus.New()

	// Set log level
	level, err := logrus.ParseLevel(config.Level)
	if err != nil {
		level = logrus.InfoLevel
	}
	logger.SetLevel(level)

	// Set formatter
	if config.Format == "json" {
		logger.SetFormatter(&logrus.JSONFormatter{
			TimestampFormat: time.RFC3339,
		})
	} else {
		logger.SetFormatter(&logrus.TextFormatter{
			FullTimestamp:   true,
			TimestampFormat: time.RFC3339,
		})
	}

	// Set output
	if config.Output == "file" && config.File != "" {
		file, err := os.OpenFile(config.File, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
		if err != nil {
			return nil, fmt.Errorf("failed to open log file: %w", err)
		}
		logger.SetOutput(file)
	}

	return &StructuredLogger{
		config:   config,
		logger:   logger,
		redactor: NewFieldRedactor(config.RedactFields),
	}, nil
}

// Context keys for correlation data
type contextKey string

const (
	CorrelationIDKey contextKey = "correlation_id"
	TraceIDKey       contextKey = "trace_id"
	SpanIDKey        contextKey = "span_id"
	UserIDKey        contextKey = "user_id"
	TenantIDKey      contextKey = "tenant_id"
	RequestIDKey     contextKey = "request_id"
	SessionIDKey     contextKey = "session_id"
)

// WithCorrelationID adds correlation ID to context
func WithCorrelationID(ctx context.Context, correlationID string) context.Context {
	if correlationID == "" {
		correlationID = uuid.New().String()
	}
	return context.WithValue(ctx, CorrelationIDKey, correlationID)
}

// WithTraceData adds trace data to context
func WithTraceData(ctx context.Context, traceID, spanID string) context.Context {
	ctx = context.WithValue(ctx, TraceIDKey, traceID)
	ctx = context.WithValue(ctx, SpanIDKey, spanID)
	return ctx
}

// WithUserData adds user data to context
func WithUserData(ctx context.Context, userID, tenantID string) context.Context {
	ctx = context.WithValue(ctx, UserIDKey, userID)
	ctx = context.WithValue(ctx, TenantIDKey, tenantID)
	return ctx
}

// WithRequestData adds request data to context
func WithRequestData(ctx context.Context, requestID, sessionID string) context.Context {
	ctx = context.WithValue(ctx, RequestIDKey, requestID)
	if sessionID != "" {
		ctx = context.WithValue(ctx, SessionIDKey, sessionID)
	}
	return ctx
}

// GetCorrelationID retrieves correlation ID from context
func GetCorrelationID(ctx context.Context) string {
	if id, ok := ctx.Value(CorrelationIDKey).(string); ok {
		return id
	}
	return ""
}

// GetTraceID retrieves trace ID from context
func GetTraceID(ctx context.Context) string {
	if id, ok := ctx.Value(TraceIDKey).(string); ok {
		return id
	}
	return ""
}

// GetSpanID retrieves span ID from context
func GetSpanID(ctx context.Context) string {
	if id, ok := ctx.Value(SpanIDKey).(string); ok {
		return id
	}
	return ""
}

// GetUserID retrieves user ID from context
func GetUserID(ctx context.Context) string {
	if id, ok := ctx.Value(UserIDKey).(string); ok {
		return id
	}
	return ""
}

// GetTenantID retrieves tenant ID from context
func GetTenantID(ctx context.Context) string {
	if id, ok := ctx.Value(TenantIDKey).(string); ok {
		return id
	}
	return ""
}

// GetRequestID retrieves request ID from context
func GetRequestID(ctx context.Context) string {
	if id, ok := ctx.Value(RequestIDKey).(string); ok {
		return id
	}
	return ""
}

// GetSessionID retrieves session ID from context
func GetSessionID(ctx context.Context) string {
	if id, ok := ctx.Value(SessionIDKey).(string); ok {
		return id
	}
	return ""
}

// createLogEntry creates a log entry from context and parameters
func (sl *StructuredLogger) createLogEntry(ctx context.Context, level LogLevel, message string, fields map[string]interface{}) *LogEntry {
	entry := &LogEntry{
		Timestamp:     time.Now(),
		Level:         level,
		Message:       message,
		Service:       sl.config.Service,
		Version:       sl.config.Version,
		CorrelationID: GetCorrelationID(ctx),
		TraceID:       GetTraceID(ctx),
		SpanID:        GetSpanID(ctx),
		UserID:        GetUserID(ctx),
		TenantID:      GetTenantID(ctx),
		RequestID:     GetRequestID(ctx),
		SessionID:     GetSessionID(ctx),
		Metadata:      make(map[string]interface{}),
	}

	// Add caller information
	if pc, file, line, ok := runtime.Caller(2); ok {
		entry.Function = runtime.FuncForPC(pc).Name()
		entry.File = file
		entry.Line = line
	}

	// Process fields
	for key, value := range fields {
		switch key {
		case "duration":
			if duration, ok := value.(time.Duration); ok {
				entry.Duration = duration
			}
		case "error":
			if err, ok := value.(error); ok {
				entry.Error = &ErrorInfo{
					Type:    fmt.Sprintf("%T", err),
					Message: err.Error(),
				}
				// Add stack trace for errors
				buf := make([]byte, 4096)
				n := runtime.Stack(buf, false)
				entry.Error.StackTrace = string(buf[:n])
			}
		case "tags":
			if tags, ok := value.([]string); ok {
				entry.Tags = tags
			}
		default:
			entry.Metadata[key] = sl.redactor.Redact(value)
		}
	}

	return entry
}

// log logs the entry using the underlying logger
func (sl *StructuredLogger) log(entry *LogEntry) {
	// Convert to logrus fields
	fields := logrus.Fields{
		"service":        entry.Service,
		"correlation_id": entry.CorrelationID,
		"trace_id":       entry.TraceID,
		"span_id":        entry.SpanID,
		"user_id":        entry.UserID,
		"tenant_id":      entry.TenantID,
		"request_id":     entry.RequestID,
		"session_id":     entry.SessionID,
		"function":       entry.Function,
		"file":           entry.File,
		"line":           entry.Line,
	}

	if entry.Duration > 0 {
		fields["duration"] = entry.Duration.String()
	}

	if entry.Error != nil {
		fields["error"] = entry.Error
	}

	if len(entry.Metadata) > 0 {
		for k, v := range entry.Metadata {
			fields[k] = v
		}
	}

	if len(entry.Tags) > 0 {
		fields["tags"] = entry.Tags
	}

	// Log with appropriate level
	switch entry.Level {
	case LevelDebug:
		sl.logger.WithFields(fields).Debug(entry.Message)
	case LevelInfo:
		sl.logger.WithFields(fields).Info(entry.Message)
	case LevelWarn:
		sl.logger.WithFields(fields).Warn(entry.Message)
	case LevelError:
		sl.logger.WithFields(fields).Error(entry.Message)
	case LevelFatal:
		sl.logger.WithFields(fields).Fatal(entry.Message)
	case LevelPanic:
		sl.logger.WithFields(fields).Panic(entry.Message)
	}
}

// Debug logs a debug message
func (sl *StructuredLogger) Debug(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelDebug, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// Info logs an info message
func (sl *StructuredLogger) Info(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelInfo, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// Warn logs a warning message
func (sl *StructuredLogger) Warn(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelWarn, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// Error logs an error message
func (sl *StructuredLogger) Error(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelError, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// Fatal logs a fatal message and exits
func (sl *StructuredLogger) Fatal(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelFatal, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// Panic logs a panic message and panics
func (sl *StructuredLogger) Panic(ctx context.Context, message string, fields ...map[string]interface{}) {
	entry := sl.createLogEntry(ctx, LevelPanic, message, sl.mergeFields(fields...))
	sl.log(entry)
}

// mergeFields merges multiple field maps
func (sl *StructuredLogger) mergeFields(fields ...map[string]interface{}) map[string]interface{} {
	merged := make(map[string]interface{})
	for _, fieldMap := range fields {
		for k, v := range fieldMap {
			merged[k] = v
		}
	}
	return merged
}

// WithFields creates a logger with pre-configured fields
func (sl *StructuredLogger) WithFields(fields map[string]interface{}) *FieldLogger {
	return &FieldLogger{
		logger: sl,
		fields: fields,
	}
}

// FieldLogger wraps StructuredLogger with pre-configured fields
type FieldLogger struct {
	logger *StructuredLogger
	fields map[string]interface{}
}

// Debug logs a debug message with pre-configured fields
func (fl *FieldLogger) Debug(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Debug(ctx, message, allFields)
}

// Info logs an info message with pre-configured fields
func (fl *FieldLogger) Info(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Info(ctx, message, allFields)
}

// Warn logs a warning message with pre-configured fields
func (fl *FieldLogger) Warn(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Warn(ctx, message, allFields)
}

// Error logs an error message with pre-configured fields
func (fl *FieldLogger) Error(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Error(ctx, message, allFields)
}

// Fatal logs a fatal message with pre-configured fields
func (fl *FieldLogger) Fatal(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Fatal(ctx, message, allFields)
}

// Panic logs a panic message with pre-configured fields
func (fl *FieldLogger) Panic(ctx context.Context, message string, fields ...map[string]interface{}) {
	allFields := fl.mergeFields(fields...)
	fl.logger.Panic(ctx, message, allFields)
}

func (fl *FieldLogger) mergeFields(fields ...map[string]interface{}) map[string]interface{} {
	merged := make(map[string]interface{})
	for k, v := range fl.fields {
		merged[k] = v
	}
	for _, fieldMap := range fields {
		for k, v := range fieldMap {
			merged[k] = v
		}
	}
	return merged
}

File: products/sdlc-cc/platform/observability/logging/python/structured_logger.py
"""
Structured logging implementation for SDLC.ai platform.
Provides JSON-formatted logging with correlation ID support and sensitive data redaction.
"""

import json
import logging
import os
import sys
import time
import traceback
import uuid
from contextlib import contextmanager
from contextvars import ContextVar
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Union
from pathlib import Path

from pythonjsonlogger import jsonlogger


class LogLevel(Enum):
    """Log levels for structured logging."""

    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"
    FATAL = "fatal"
    CRITICAL = "critical"


class ErrorInfo:
    """Detailed error information for structured logging."""

    def __init__(
        self,
        error_type: str,
        message: str,
        stack_trace: Optional[str] = None,
        cause: Optional[str] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        self.error_type = error_type
        self.message = message
        self.stack_trace = stack_trace
        self.cause = cause
        self.details = details or {}


class FieldRedactor:
    """Handles sensitive data redaction in log fields."""

    def __init__(self, sensitive_fields: Optional[List[str]] = None):
        self.sensitive_fields = set(field.lower() for field in (sensitive_fields or []))
        self.default_sensitive = {
            "password",
            "token",
            "secret",
            "key",
            "auth",
            "credential",
            "bearer",
            "api_key",
            "access_token",
        }
        self.sensitive_fields.update(self.default_sensitive)

    def redact(self, data: Any) -> Any:
        """Recursively redact sensitive fields from data."""
        if isinstance(data, dict):
            redacted = {}
            for key, value in data.items():
                if self._is_sensitive(key):
                    redacted[key] = "***REDACTED***"
                else:
                    redacted[key] = self.redact(value)
            return redacted
        elif isinstance(data, list):
            return [self.redact(item) for item in data]
        else:
            return data

    def _is_sensitive(self, field_name: str) -> bool:
        """Check if a field name indicates sensitive data."""
        field_lower = field_name.lower()
        return any(sensitive in field_lower for sensitive in self.sensitive_fields)


class LogEntry:
    """Structured log entry with correlation support."""

    def __init__(
        self,
        level: LogLevel,
        message: str,
        service: str,
        version: Optional[str] = None,
        timestamp: Optional[datetime] = None,
        trace_id: Optional[str] = None,
        span_id: Optional[str] = None,
        correlation_id: Optional[str] = None,
        user_id: Optional[str] = None,
        tenant_id: Optional[str] = None,
        request_id: Optional[str] = None,
        session_id: Optional[str] = None,
        source: Optional[str] = None,
        function: Optional[str] = None,
        file: Optional[str] = None,
        line: Optional[int] = None,
        duration: Optional[float] = None,
        error: Optional[ErrorInfo] = None,
        metadata: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None,
    ):
        self.timestamp = timestamp or datetime.utcnow()
        self.level = level
        self.message = message
        self.service = service
        self.version = version
        self.trace_id = trace_id
        self.span_id = span_id
        self.correlation_id = correlation_id
        self.user_id = user_id
        self.tenant_id = tenant_id
        self.request_id = request_id
        self.session_id = session_id
        self.source = source
        self.function = function
        self.file = file
        self.line = line
        self.duration = duration
        self.error = error
        self.metadata = metadata or {}
        self.tags = tags or []

    def to_dict(self) -> Dict[str, Any]:
        """Convert log entry to dictionary for JSON serialization."""
        result = {
            "timestamp": self.timestamp.isoformat(),
            "level": self.level.value,
            "message": self.message,
            "service": self.service,
            "correlation_id": self.correlation_id,
            "trace_id": self.trace_id,
            "span_id": self.span_id,
            "user_id": self.user_id,
            "tenant_id": self.tenant_id,
            "request_id": self.request_id,
            "session_id": self.session_id,
            "metadata": self.metadata,
            "tags": self.tags,
        }

        # Add optional fields only if they have values
        if self.version:
            result["version"] = self.version
        if self.source:
            result["source"] = self.source
        if self.function:
            result["function"] = self.function
        if self.file:
            result["file"] = self.file
        if self.line:
            result["line"] = self.line
        if self.duration:
            result["duration"] = self.duration
        if self.error:
            result["error"] = {
                "type": self.error.error_type,
                "message": self.error.message,
                "stack_trace": self.error.stack_trace,
                "cause": self.error.cause,
                "details": self.error.details,
            }

        return result


# Context variables for correlation data
correlation_id_var: ContextVar[Optional[str]] = ContextVar(
    "correlation_id", default=None
)
trace_id_var: ContextVar[Optional[str]] = ContextVar("trace_id", default=None)
span_id_var: ContextVar[Optional[str]] = ContextVar("span_id", default=None)
user_id_var: ContextVar[Optional[str]] = ContextVar("user_id", default=None)
tenant_id_var: ContextVar[Optional[str]] = ContextVar("tenant_id", default=None)
request_id_var: ContextVar[Optional[str]] = ContextVar("request_id", default=None)
session_id_var: ContextVar[Optional[str]] = ContextVar("session_id", default=None)


class LoggerConfig:
    """Configuration for structured logger."""

    def __init__(
        self,
        service: str,
        version: str = "1.0.0",
        environment: str = "development",
        level: str = "info",
        format_type: str = "json",
        output: str = "stdout",
        file_path: Optional[str] = None,
        max_size_mb: int = 100,
        max_backups: int = 5,
        max_age_days: int = 30,
        compress: bool = True,
        enable_tracing: bool = True,
        redact_fields: Optional[List[str]] = None,
    ):
        self.service = service
        self.version = version
        self.environment = environment
        self.level = level
        self.format_type = format_type
        self.output = output
        self.file_path = file_path
        self.max_size_mb = max_size_mb
        self.max_backups = max_backups
        self.max_age_days = max_age_days
        self.compress = compress
        self.enable_tracing = enable_tracing
        self.redact_fields = redact_fields or []

    @classmethod
    def default(cls) -> "LoggerConfig":
        """Create default logger configuration."""
        return cls(
            service="sdlc-platform",
            version="1.0.0",
            environment="development",
            level="info",
            format_type="json",
            output="stdout",
            max_size_mb=100,
            max_backups=5,
            max_age_days=30,
            compress=True,
            enable_tracing=True,
            redact_fields=[
                "password",
                "token",
                "secret",
                "key",
                "auth",
                "credential",
                "bearer",
                "api_key",
                "access_token",
            ],
        )


class StructuredLogger:
    """Structured logger with correlation support and sensitive data redaction."""

    def __init__(self, config: LoggerConfig):
        self.config = config
        self.redactor = FieldRedactor(config.redact_fields)
        self._setup_logger()

    def _setup_logger(self):
        """Setup the underlying Python logger."""
        self.logger = logging.getLogger(f"{self.config.service}-structured")
        self.logger.setLevel(getattr(logging, self.config.level.upper()))

        # Clear existing handlers
        self.logger.handlers.clear()

        # Create formatter
        if self.config.format_type == "json":
            formatter = jsonlogger.JsonFormatter(
                "%(asctime)s %(name)s %(levelname)s %(message)s"
            )
        else:
            formatter = logging.Formatter(
                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
            )

        # Create handler
        if self.config.output == "file" and self.config.file_path:
            # Ensure log directory exists
            log_dir = Path(self.config.file_path).parent
            log_dir.mkdir(parents=True, exist_ok=True)

            from logging.handlers import RotatingFileHandler

            handler = RotatingFileHandler(
                self.config.file_path,
                maxBytes=self.config.max_size_mb * 1024 * 1024,
                backupCount=self.config.max_backups,
            )
        else:
            handler = logging.StreamHandler(sys.stdout)

        handler.setFormatter(formatter)
        self.logger.addHandler(handler)

    def _get_caller_info(self) -> tuple[Optional[str], Optional[str], Optional[int]]:
        """Get caller information for logging."""
        import inspect

        frame = inspect.currentframe()
        try:
            # Go up the call stack to find the actual caller
            caller_frame = frame.f_back.f_back
            if caller_frame:
                function = caller_frame.f_code.co_name
                file = caller_frame.f_code.co_filename
                line = caller_frame.f_lineno
                return function, file, line
        finally:
            del frame
        return None, None, None

    def _create_log_entry(self, level: LogLevel, message: str, **kwargs) -> LogEntry:
        """Create a log entry with correlation data from context."""
        function, file, line = self._get_caller_info()

        return LogEntry(
            level=level,
            message=message,
            service=self.config.service,
            version=self.config.version,
            trace_id=trace_id_var.get(),
            span_id=span_id_var.get(),
            correlation_id=correlation_id_var.get(),
            user_id=user_id_var.get(),
            tenant_id=tenant_id_var.get(),
            request_id=request_id_var.get(),
            session_id=session_id_var.get(),
            function=function,
            file=file,
            line=line,
            **kwargs,
        )

    def _log(self, entry: LogEntry):
        """Log the entry using the underlying logger."""
        log_data = entry.to_dict()

        # Apply redaction to metadata
        if entry.metadata:
            log_data["metadata"] = self.redactor.redact(entry.metadata)

        # Use appropriate logging level
        if entry.level == LogLevel.DEBUG:
            self.logger.debug(json.dumps(log_data))
        elif entry.level == LogLevel.INFO:
            self.logger.info(json.dumps(log_data))
        elif entry.level == LogLevel.WARN:
            self.logger.warning(json.dumps(log_data))
        elif entry.level == LogLevel.ERROR:
            self.logger.error(json.dumps(log_data))
        elif entry.level == LogLevel.FATAL or entry.level == LogLevel.CRITICAL:
            self.logger.critical(json.dumps(log_data))

    def debug(self, message: str, **kwargs):
        """Log a debug message."""
        entry = self._create_log_entry(LogLevel.DEBUG, message, **kwargs)
        self._log(entry)

    def info(self, message: str, **kwargs):
        """Log an info message."""
        entry = self._create_log_entry(LogLevel.INFO, message, **kwargs)
        self._log(entry)

    def warn(self, message: str, **kwargs):
        """Log a warning message."""
        entry = self._create_log_entry(LogLevel.WARN, message, **kwargs)
        self._log(entry)

    def error(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log an error message."""
        error_info = None
        if error:
            error_info = ErrorInfo(
                error_type=type(error).__name__,
                message=str(error),
                stack_trace=traceback.format_exc(),
                details=kwargs.pop("error_details", None),
            )

        entry = self._create_log_entry(
            LogLevel.ERROR, message, error=error_info, **kwargs
        )
        self._log(entry)

    def fatal(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log a fatal message."""
        error_info = None
        if error:
            error_info = ErrorInfo(
                error_type=type(error).__name__,
                message=str(error),
                stack_trace=traceback.format_exc(),
                details=kwargs.pop("error_details", None),
            )

        entry = self._create_log_entry(
            LogLevel.FATAL, message, error=error_info, **kwargs
        )
        self._log(entry)

    def critical(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log a critical message."""
        error_info = None
        if error:
            error_info = ErrorInfo(
                error_type=type(error).__name__,
                message=str(error),
                stack_trace=traceback.format_exc(),
                details=kwargs.pop("error_details", None),
            )

        entry = self._create_log_entry(
            LogLevel.CRITICAL, message, error=error_info, **kwargs
        )
        self._log(entry)

    def with_fields(self, **fields) -> "FieldLogger":
        """Create a logger with pre-configured fields."""
        return FieldLogger(self, fields)


class FieldLogger:
    """Logger wrapper with pre-configured fields."""

    def __init__(self, logger: StructuredLogger, fields: Dict[str, Any]):
        self.logger = logger
        self.fields = fields

    def _merge_fields(self, **kwargs) -> Dict[str, Any]:
        """Merge pre-configured fields with call-time fields."""
        merged = self.fields.copy()
        merged.update(kwargs)
        return merged

    def debug(self, message: str, **kwargs):
        """Log a debug message with pre-configured fields."""
        self.logger.debug(message, **self._merge_fields(**kwargs))

    def info(self, message: str, **kwargs):
        """Log an info message with pre-configured fields."""
        self.logger.info(message, **self._merge_fields(**kwargs))

    def warn(self, message: str, **kwargs):
        """Log a warning message with pre-configured fields."""
        self.logger.warn(message, **self._merge_fields(**kwargs))

    def error(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log an error message with pre-configured fields."""
        self.logger.error(message, error=error, **self._merge_fields(**kwargs))

    def fatal(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log a fatal message with pre-configured fields."""
        self.logger.fatal(message, error=error, **self._merge_fields(**kwargs))

    def critical(self, message: str, error: Optional[Exception] = None, **kwargs):
        """Log a critical message with pre-configured fields."""
        self.logger.critical(message, error=error, **self._merge_fields(**kwargs))


# Context management functions
def with_correlation_id(correlation_id: Optional[str] = None) -> ContextVar:
    """Set correlation ID in context."""
    if correlation_id is None:
        correlation_id = str(uuid.uuid4())
    return correlation_id_var.set(correlation_id)


def with_trace_data(trace_id: str, span_id: str) -> tuple[ContextVar, ContextVar]:
    """Set trace data in context."""
    return trace_id_var.set(trace_id), span_id_var.set(span_id)


def with_user_data(user_id: str, tenant_id: str) -> tuple[ContextVar, ContextVar]:
    """Set user data in context."""
    return user_id_var.set(user_id), tenant_id_var.set(tenant_id)


def with_request_data(
    request_id: str, session_id: Optional[str] = None
) -> tuple[ContextVar, Optional[ContextVar]]:
    """Set request data in context."""
    request_token = request_id_var.set(request_id)
    session_token = session_id_var.set(session_id) if session_id else None
    return request_token, session_token


# Context manager for logging operations
@contextmanager
def log_context(
    correlation_id: Optional[str] = None,
    trace_id: Optional[str] = None,
    span_id: Optional[str] = None,
    user_id: Optional[str] = None,
    tenant_id: Optional[str] = None,
    request_id: Optional[str] = None,
    session_id: Optional[str] = None,
):
    """Context manager for setting logging context."""
    tokens = []

    try:
        if correlation_id:
            tokens.append(with_correlation_id(correlation_id))
        if trace_id and span_id:
            trace_token, span_token = with_trace_data(trace_id, span_id)
            tokens.extend([trace_token, span_token])
        if user_id and tenant_id:
            user_token, tenant_token = with_user_data(user_id, tenant_id)
            tokens.extend([user_token, tenant_token])
        if request_id:
            request_token, session_token = with_request_data(request_id, session_id)
            tokens.append(request_token)
            if session_token:
                tokens.append(session_token)

        yield

    finally:
        # Reset context variables
        for token in tokens:
            if hasattr(token, "reset"):
                token.reset()


# Global logger instance
_global_logger: Optional[StructuredLogger] = None


def get_logger() -> StructuredLogger:
    """Get the global structured logger instance."""
    global _global_logger
    if _global_logger is None:
        _global_logger = StructuredLogger(LoggerConfig.default())
    return _global_logger


def configure_logger(config: LoggerConfig):
    """Configure the global structured logger."""
    global _global_logger
    _global_logger = StructuredLogger(config)


# Convenience functions using global logger
def debug(message: str, **kwargs):
    """Log a debug message using the global logger."""
    get_logger().debug(message, **kwargs)


def info(message: str, **kwargs):
    """Log an info message using the global logger."""
    get_logger().info(message, **kwargs)


def warn(message: str, **kwargs):
    """Log a warning message using the global logger."""
    get_logger().warn(message, **kwargs)


def error(message: str, error: Optional[Exception] = None, **kwargs):
    """Log an error message using the global logger."""
    get_logger().error(message, error=error, **kwargs)


def fatal(message: str, error: Optional[Exception] = None, **kwargs):
    """Log a fatal message using the global logger."""
    get_logger().fatal(message, error=error, **kwargs)


def critical(message: str, error: Optional[Exception] = None, **kwargs):
    """Log a critical message using the global logger."""
    get_logger().critical(message, error=error, **kwargs)
